consensus/src/payload_manager/quorum_store_payload_manager.rs

```diff
@@ -113,49 +68,53 @@ pub struct QuorumStorePayloadManager {
 impl QuorumStorePayloadManager {
     pub fn new(
         batch_reader: Arc<dyn BatchReader>,
-        coordinator_tx: Sender<CoordinatorCommand>,
+        commit_notifier: Box<dyn TQuorumStoreCommitNotifier>,
         maybe_consensus_publisher: Option<Arc<ConsensusPublisher>>,
         ordered_authors: Vec<PeerId>,
         address_to_validator_index: HashMap<PeerId, usize>,
     ) -> Self {
         Self {
             batch_reader,
-            coordinator_tx,
+            commit_notifier,
             maybe_consensus_publisher,
             ordered_authors,
             address_to_validator_index,
         }
     }
 
     fn request_transactions(
-        batches: Vec<(BatchInfo, Arc<Mutex<Vec<PeerId>>>)>,
+        batches: Vec<(BatchInfo, Vec<PeerId>)>,
         block_timestamp: u64,
         batch_reader: Arc<dyn BatchReader>,
-    ) -> Vec<(
-        HashValue,
-        oneshot::Receiver<ExecutorResult<Vec<SignedTransaction>>>,
-    )> {
-        let mut receivers = Vec::new();
+    ) -> Vec<Shared<Pin<Box<dyn Future<Output = ExecutorResult<Vec<SignedTransaction>>> + Send>>>>
+    {
+        let mut futures = Vec::new();
         for (batch_info, responders) in batches {
             trace!(
                 "QSE: requesting batch {:?}, time = {}",
                 batch_info,
                 block_timestamp
             );
             if block_timestamp <= batch_info.expiration() {
-                receivers.push((
-                    *batch_info.digest(),
-                    batch_reader.get_batch(
-                        *batch_info.digest(),
-                        batch_info.expiration(),
-                        responders.clone(),
-                    ),
-                ));
+                futures.push(batch_reader.get_batch(batch_info, responders.clone()));
             } else {
                 debug!("QSE: skipped expired batch {}", batch_info.digest());
             }
         }
-        receivers
+        futures
+    }
+
+    async fn request_and_wait_transactions(
+        batches: Vec<(BatchInfo, Vec<PeerId>)>,
+        block_timestamp: u64,
+        batch_reader: Arc<dyn BatchReader>,
+    ) -> ExecutorResult<Vec<SignedTransaction>> {
+        let futures = Self::request_transactions(batches, block_timestamp, batch_reader);
+        let mut all_txns = Vec::new();
+        for result in futures::future::join_all(futures).await {
```

Comment:
what's the difference between join_all vs just iterate through individual futures and await?

Reply:
I hope join_all drive multiple futures at the same time.